crates/kitsune2/src/factories/mem_peer_store.rs

```diff
@@ -0,0 +1,261 @@
+//! A production-ready memory-based peer store.
+
+use kitsune2_api::{agent::*, config::*, peer_store::*, *};
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
+
+/// Configuration parameters for [MemPeerStoreFactory]
+#[derive(Debug, Default, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MemPeerStoreConfig {}
+
+impl ModConfig for MemPeerStoreConfig {}
+
+/// A production-ready memory-based peer store factory.
+#[derive(Debug)]
+pub struct MemPeerStoreFactory {}
+
+impl MemPeerStoreFactory {
+    /// Construct a new MemPeerStoreFactory
+    pub fn create() -> DynPeerStoreFactory {
+        let out: DynPeerStoreFactory = Arc::new(Self {});
+        out
+    }
+}
+
+impl PeerStoreFactory for MemPeerStoreFactory {
+    fn default_config(&self, _config: &mut Config) {}
+
+    fn create(
+        &self,
+        _builder: Arc<builder::Builder>,
+    ) -> BoxFut<'static, K2Result<DynPeerStore>> {
+        Box::pin(async move {
+            let out: DynPeerStore = Arc::new(MemPeerStore::new());
+            Ok(out)
+        })
+    }
+}
+
+struct MemPeerStore(Mutex<Inner>);
+
+impl std::fmt::Debug for MemPeerStore {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("MemPeerStore").finish()
+    }
+}
+
+impl MemPeerStore {
+    pub fn new() -> Self {
+        Self(Mutex::new(Inner::new(std::time::Instant::now())))
+    }
+}
+
+impl peer_store::PeerStore for MemPeerStore {
+    fn insert(
```

Comment:
So it looks like we keep all peers unless they are expired right?

Does it make sense to follow a similar approach to the bootstrap server of splitting the store between long-running and recently added nodes? And to limit the number of peers we store? Seems like the same rationale applies unless I'm missing something?

Reply:
Mmm, the selecting the peers is a higher level concern, but making them available is a this-level concern.

Basically, anyone we overlap storage arcs with we need to keep here, because we must gossip with all of them. We *could* drop agents outside that set, but I'd consider that a premature optimization, because it's actually helpful to have them in our store if we need to publish to or get from that neighborhood. We can get more fancy later about cleaning up the peer store if we find it's using too much memory.